osu.Game/Rulesets/UI/Playfield.cs

```diff
@@ -336,6 +343,29 @@ DrawableHitObject IPooledHitObjectProvider.GetPooledDrawableRepresentation(HitOb
             });
         }
 
+        private readonly Dictionary<ISampleInfo, DrawablePool<PoolableSkinnableSample>> samplePools = new Dictionary<ISampleInfo, DrawablePool<PoolableSkinnableSample>>();
+
+        public PoolableSkinnableSample GetPooledSample(ISampleInfo sampleInfo)
+        {
+            if (!samplePools.TryGetValue(sampleInfo, out var existingPool))
+                AddInternal(samplePools[sampleInfo] = existingPool = new DrawableSamplePool(sampleInfo, 1));
```

Comment:
slightly worried about the thread safety of this. sample changes after initial load i believe *should* be relatively safe because they should be occurring on the update thread, but how about the initial load? i believe the flow will be something like

```
DrawableHitObject.LoadCompleteAsync()
DrawableHitObject.Apply(HitObject, HitObjectLifetimeEntry)
DrawableHitObject.onSamplesChanged(object, NotifyCollectionChangedEventArgs)
DrawableHitObject.LoadSamples()
SkinnableSound.Samples.set
SkinnableSound.updateSamples()
Playfield.GetPooledSample(ISampleInfo) (via interface)
```

which i *think* is outside of the update loop. what i'm worried about is silently creating two pools due to lack of thread safety of the dictionary, but i'm not sure whether i'm being paranoid or not.

Reply:
I don't believe we ever load `DrawableHitObjects` asynchronously (correct me if i'm wrong on this one), so maybe a non-issue right now, but still something worth considering.

At very least, may want to add a `ThreadSafety.EnsureUpdateThread` call here.

Reply:
I think this is fine until it comes up as an issue (crash). The scenario explained by @bdach is a non-issue because at the point of initial load the player hierarchy has not been committed to a root, so these operations are allowed. Also, it should be running the full load sequence on a single thread.